.github/workflows/ci.yml

```diff
@@ -0,0 +1,166 @@
+            cache_scope: rust
+            build-args: |
+              PACKAGE=relay
+    needs:
+      - rust
+      - elixir
+      - static-analysis
+    permissions:
+      contents: read
+      id-token: write
+    env:
+      # mark:automatic-version
+      VERSION: "1.20231001.0"
+      APPLICATION_NAME: ${{ matrix.image_name }}
+      REGISTRY: us-east1-docker.pkg.dev
+      GCLOUD_PROJECT: firezone-staging
+      GOOGLE_CLOUD_PROJECT: firezone-staging
+      CLOUDSDK_PROJECT: firezone-staging
+      CLOUDSDK_CORE_PROJECT: firezone-staging
+      GCP_PROJECT: firezone-staging
+    steps:
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      - uses: actions/checkout@v4
+      - id: auth
+        uses: google-github-actions/auth@v1
+        with:
+          token_format: access_token
+          workload_identity_provider: "projects/397012414171/locations/global/workloadIdentityPools/github-actions/providers/github-actions"
+          service_account: "github-actions@github-iam-387915.iam.gserviceaccount.com"
+          export_environment_variables: false
+      - name: Change current gcloud account
+        run: gcloud --quiet config set project ${GCLOUD_PROJECT}
+      - name: Login to Google Artifact Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: oauth2accesstoken
+          password: ${{ steps.auth.outputs.access_token }}
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v5
+        with:
+          platforms: linux/amd64
+          build-args: ${{ matrix.build-args }}
+          context: ${{ matrix.cache_scope }}/
+          cache-from: type=gha,scope=${{ matrix.cache_scope }}
+          cache-to:
+            ${{ github.ref == 'refs/heads/main' &&
+            format('type=gha,mode=max,scope={0}', matrix.scope) || '' }}
+          file: ${{ matrix.cache_scope }}/Dockerfile
+          push: ${{ github.ref == 'refs/head/main' }}
+          tags: |
+            ${{ env.REGISTRY }}/${{ env.GCLOUD_PROJECT }}/firezone/${{ matrix.image_name }}:${{ env.VERSION }}
+            ${{ env.REGISTRY }}/${{ env.GCLOUD_PROJECT }}/firezone/${{ matrix.image_name }}:${{ github.sha }}
+          outputs: type=docker,dest=/tmp/${{ matrix.image_name }}.tar
+      - name: Upload built images as artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: images
+          path: /tmp/${{ matrix.image_name }}.tar
+
+  integration-tests:
+    needs: build-images
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        include:
+          - test_name: Relayed flow
+            setup: |
+              #  Disallow traffic between gateway and client container
+              sudo iptables -I FORWARD 1 -s  172.28.0.100 -d 172.28.0.105 -j DROP
+              sudo iptables -I FORWARD 1 -s  172.28.0.105 -d 172.28.0.100 -j DROP
```

Comment:
We can also do `docker exec` to simulate different kinds of NATs:

Full Cone NAT:
```
iptables -t nat -A POSTROUTING -o eth1 -j SNAT --to-source "public IP"
iptables -t nat -A PREROUTING -i eth1 -j DNAT --to-destination "private IP"
```
Restricted Cone NAT:
```
iptables -t nat -A POSTROUTING -o eth1 -p udp -j SNAT --to-source "public IP"
iptables -t nat -A PREROUTING -i eth1 -p udp -j DNAT --to-destination "private IP"
iptables -A INPUT -i eth1 -p udp -m state --state ESTABLISHED,RELATED -j ACCEPT
iptables -A INPUT -i eth1 -p udp -m state --state NEW -j DROP
```
Port Restricted Cone NAT:
```
iptables -t nat -A POSTROUTING -o eth0 -j SNAT --to-source "public IP"
```
Symmetric NAT:
```
echo "1" >/proc/sys/net/ipv4/ip_forward
iptables --flush
iptables -t nat -A POSTROUTING -o eth1 -j MASQUERADE --random
iptables -A FORWARD -i eth1 -o eth0 -m state --state RELATED,ESTABLISHED -j ACCEPT
iptables -A FORWARD -i eth0 -o eth1 -j ACCEPT
```

Reply:
I'll save it for another PR. Refs #2028